File: Stack/C programs/factorial.c
// Write a C program to calculate factorial with the help of stack.

#include <stdio.h>
#define MAX 2
void push(int st[], int *, int);
int pop(int st[], int *);

int main()
{
    int st[MAX], top = -1, n, temp;
    printf("Enter a Number: ");
    scanf("%d", &n);
    push(st, &top, 1);
    for (int i = 2; i <= n; i++)
    {
        temp = pop(st, &top);
        push(st, &top, temp * i);
    }
    printf("Factorial: %d", pop(st, &top));
    return 0;
}

void push(int st[], int *top, int val)
{
    if (*top == MAX - 1)
        printf("Stack Overflow!");
    else
        st[++(*top)] = val;
}

int pop(int st[], int *top)
{
    if (*top == -1)
    {
        printf("Stack Underflow!");
        return -1;
    }
    else
        return st[(*top)--];
}

// OUTPUT 1
// Enter a Number: 0
// Factorial: 1

// OUTPUT 2
// Enter a Number: 5
// Factorial: 120

File: Stack/Python Programs/factorial.py
# Write a program to calculate factorial with the help of stack.

stack1 = []


def push(item):
    stack1.append(item)
    return stack1


def pop():
    if len(stack1) == 0:
        return "Stack is empty"
    else:
        return stack1.pop()


if __name__ == '__main__':
    n = int(input("Enter a Number: "))
    push(1)
    for i in range(2, n+1):
        temp = pop()
        push(temp*i)
    print("Factorial: "+str(pop()))


# OUTPUT
# Enter a Number: 6
# Factorial: 720
